provider/infoblox.go

```diff
@@ -0,0 +1,389 @@
+
+	transportConfig := ibclient.NewTransportConfig(
+		strconv.FormatBool(sslVerify),
+		int(httpRequestTimeout),
+		int(httpPoolConnections),
+	)
+
+	requestBuilder := &ibclient.WapiRequestBuilder{}
+	requestor := &ibclient.WapiHttpRequestor{}
+
+	client, err := ibclient.NewConnector(hostConfig, transportConfig, requestBuilder, requestor)
+
+	if err != nil {
+		return nil, err
+	}
+
+	provider := &InfobloxProvider{
+		client:       client,
+		domainFilter: domainFilter,
+		dryRun:       dryRun,
+	}
+
+	return provider, nil
+}
+
+// Records gets the current records.
+func (p *InfobloxProvider) Records() (endpoints []*endpoint.Endpoint, err error) {
+	zones, err := p.zones()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, zone := range zones {
+		var resA []ibclient.RecordA
+		objA := ibclient.NewRecordA(
+			ibclient.RecordA{
+				Zone: zone.Fqdn,
+			},
+		)
+		err = p.client.GetObject(objA, "", &resA)
+		if err != nil {
+			return nil, err
+		}
+		for _, res := range resA {
+			endpoints = append(endpoints, endpoint.NewEndpoint(res.Name, res.Ipv4Addr, endpoint.RecordTypeA))
+		}
+
+		var resC []ibclient.RecordCNAME
+		objC := ibclient.NewRecordCNAME(
+			ibclient.RecordCNAME{
+				Zone: zone.Fqdn,
+			},
+		)
+		err = p.client.GetObject(objC, "", &resC)
+		if err != nil {
+			return nil, err
+		}
+		for _, res := range resC {
+			endpoints = append(endpoints, endpoint.NewEndpoint(res.Name, res.Canonical, endpoint.RecordTypeCNAME))
+		}
+
+		var resT []ibclient.RecordTXT
+		objT := ibclient.NewRecordTXT(
+			ibclient.RecordTXT{
+				Zone: zone.Fqdn,
+			},
+		)
+		err = p.client.GetObject(objT, "", &resT)
+		if err != nil {
+			return nil, err
+		}
+		for _, res := range resT {
+			if _, err := strconv.Unquote(res.Text); err != nil {
+				res.Text = strconv.Quote(res.Text)
+			}
+			endpoints = append(endpoints, endpoint.NewEndpoint(res.Name, res.Text, endpoint.RecordTypeTXT))
+		}
+	}
+	return endpoints, nil
+}
+
+// ApplyChanges applies the given changes.
+func (p *InfobloxProvider) ApplyChanges(changes *plan.Changes) error {
+	zones, err := p.zones()
+	if err != nil {
+		return err
+	}
+
+	created, deleted, updated := p.mapChanges(zones, changes)
+	p.createRecords(created)
+	p.deleteRecords(deleted)
+	p.updateRecords(updated)
+	return nil
+}
+
+func (p *InfobloxProvider) zones() ([]ibclient.ZoneAuth, error) {
+	var res, result []ibclient.ZoneAuth
+	obj := ibclient.NewZoneAuth(ibclient.ZoneAuth{})
+	err := p.client.GetObject(obj, "", &res)
+
+	if err != nil {
+		return nil, err
+	}
+
+	for _, zone := range res {
+		if p.domainFilter.Match(zone.Fqdn) {
+			result = append(result, zone)
+		}
+	}
+
+	return result, nil
+}
+
+type infobloxChangeMap map[string][]*endpoint.Endpoint
+
+func (p *InfobloxProvider) mapChanges(zones []ibclient.ZoneAuth, changes *plan.Changes) (infobloxChangeMap, infobloxChangeMap, infobloxChangeMap) {
+	ignored := map[string]bool{}
```

Comment:
what is the use of `ignored` map ?

Reply:
I borrowed the logic from the Azure provider. It looks like it simply stores the state of a particular endpoint's membership to a domain for the purposes of suppressing duplicate log messages.

Reply:
I don't see having duplicated logs as that much of a problem to be honest, I am fine with `ignored` map being there though 👍